Initial setup

aroelo opened the pull request.

Commit:
Initial commit

Commit:
Use feature branch for unit test

Commit:
Small fixes

Commit:
Update github action workflow

Commit:
Add strategy.ini file

Commit:
Build image

Commit:
Build image

Commit:
Build image

Commit:
Build image

Commit:
Small fixes

Commit:
Test terraform workflow

Commit:
Test terraform workflow

Commit:
Test terraform workflow

Commit:
Test terraform workflow

Commit:
Reformat and make some variables sensitive

Commit:
Mask gcp project id

Commit:
Mask gcp project id

Commit:
Mask gcp project id

Commit:
Test

Commit:
Run workflow with PR

Commit:
Run workflow with PR

Commit:
update

Commit:
update

Commit:
update

Commit:
update

Commit:
update

Commit:
update

Commit:
Try coki-api-base from pypi

Commit:
Update documentation

Commit:
Reformat terraform configuration

Commit:
Update workflow

Commit:
Update license dates

Commit:
Move create_app to coki-api-base

Commit:
Add docs

Commit:
Add docs

Commit:
Add docs

Commit:
Add docs, after fixing CRLF line separators

Commit:
Add requirements.txt

Commit:
Add requirements.txt

Commit:
Add requirements.txt

Commit:
Add tests to check if openapi file is valid

Commit:
Upgrade Jinja2 version

Commit:
Upgrade flask version

Commit:
Build Dockerfile using coki-api-base

Commit:
Use test pypi coki-api-base for unit tests

Commit:
Use openapi template blocks

Commit:
Move installation docs

Commit:
Update openapi template

Commit:
Re-run openapi generator

jdddog commented on academic_observatory_api/openapi.yaml.jinja2:
```diff
@@ -0,0 +1,313 @@
+swagger: '2.0'
+info:
+  title: academic-observatory-workflows API
```
```suggestion
  title: Academic Observatory API
```

jdddog commented on docs/api/api_rest.rst:
```diff
@@ -0,0 +1,6 @@
+REST API Specification
+----------------------------
+The REST API specification for managing and accessing data from the Observatory Platform.
```
Description should be updated

jdddog commented on docs/api/openapi.yaml:
```diff
@@ -0,0 +1,309 @@
+swagger: '2.0'
+info:
+  title: academic-observatory-workflows API
```
Also, copy the updated openapi.yaml here when it is ready.

jdddog commented on academic_observatory_api/client/api/observatory_api.py:
```diff
@@ -0,0 +1,857 @@
+from academic_observatory_api.client.model.query_response import QueryResponse
+
+
+class ObservatoryApi(object):
```
I think it would be better if this was called `AcademicObservatoryApi`, I presume that something will need to be updated in the openapi.yaml file

jdddog commented on terraform/main.tf:
```diff
@@ -0,0 +1,76 @@
+terraform {
+  backend "remote" {
+    workspaces {
+      prefix = "observatory-"
```
Won't this be deployed from a different workspace such as `academic-observatory-api-`?

aroelo replied:
I've changed this to use the 'cloud' configuration instead like you commented yesterday. I've set it up so that it will use a tag 'academic-observatory-api', this is instead of the prefix.

jdddog commented on terraform/main.tf:
```diff
@@ -0,0 +1,76 @@
+}
+
+# Get info from the observatory workspace if this is given
+data "terraform_remote_state" "observatory" {
+  count   = var.api_type.type == "observatory_api" ? 1 : 0
+  backend = "remote"
+  config = {
+    organization = var.api_type.observatory_organization
+    workspaces = {
+      name = var.api_type.observatory_workspace
+    }
+  }
+}
+
+locals {
+  # Set the environment variables for the Cloud Run backend
+  env_vars = (
+    var.api_type.type == "observatory_api" ?
+    tomap({
+      "OBSERVATORY_DB_URI" = data.terraform_remote_state.observatory[0].outputs.observatory_db_uri
+    }) :
+    tomap({
+      "ES_HOST"    = var.api_type.elasticsearch_host,
+      "ES_API_KEY" = var.api_type.elasticsearch_api_key,
+    })
+  )
+
+  # Set the annotations for the cloud run backend.
+  cloud_run_annotations = (
+    var.api_type.type == "observatory_api" ?
+    tomap({
+      "autoscaling.knative.dev/maxScale"        = "10"
+      "run.googleapis.com/vpc-access-egress"    = "private-ranges-only"
+      "run.googleapis.com/vpc-access-connector" = "projects/${var.google_cloud.project_id}/locations/${var.google_cloud.region}/connectors/${data.terraform_remote_state.observatory[0].outputs.vpc_connector_name}"
+    }) :
+    tomap({
+      "autoscaling.knative.dev/maxScale" = "10"
+    })
+  )
+}
```
You should be able to remove all of the references to the Observatory API as this project just relates to the Academic Observatory API.

The `vpc_connector_name` could just be passed in as a Terraform variable.

jdddog commented on terraform/variables.tf:
```diff
@@ -0,0 +1,69 @@
+  type        = string
+}
+
+variable "api_type" {
+  description = <<EOF
+Setting related to the specific api type, either the data api or observatory api.
+The data api requires the observatory organization and workspace set, while the data api requires the elasticsearch
+host and api key set.
+EOF
+  type = object({
+    type                     = string
+    observatory_organization = string
+    observatory_workspace    = string
+    elasticsearch_api_key    = string
+    elasticsearch_host       = string
+  })
+  sensitive = true
+
+  validation {
+    condition     = var.api_type.type == "data_api" || var.api_type.type == "observatory_api"
+    error_message = "The api type must either be 'data_api' or 'observatory_api'."
+  }
+  validation {
+    condition = (
+      var.api_type.type == "data_api" && var.api_type.elasticsearch_host != "" && var.api_type.elasticsearch_api_key != "" ||
+      var.api_type.type == "observatory_api" && var.api_type.observatory_organization != "" && var.api_type.observatory_workspace != ""
+    )
+    error_message = "Elasticsearch host and api key can not be empty when the api type is set to 'data_api' and observatory organization and workspace can not be empty when the api type is set to 'observatory_api'.."
+  }
+}
```
You should be able to remove `type`, `observatory_organization` and `observatory_workspace` as these are not necessary for deploying the Academic Observatory API.

elasticsearch_host and elasticsearch_api_key could be standalone Terraform variables.

jdddog commented on academic_observatory_api/openapi.yaml.jinja2:
```diff
@@ -0,0 +1,313 @@
+{# Set security and host settings #}
+{% include 'openapi_blocks.yaml.jinja2' %}
+
+{# Tag settings #}
+{% if type == 'openapi_generator' %}
+{% set reports_tag="Observatory" %}
+tags:
+- name: Observatory
+  description: the Observatory API
+{% else %}
+{% set reports_tag="Reports" %}
+tags:
+- name: Reports
+  description: Test query data to build Open Access reports.
+{% endif %}
```
```suggestion
{# Tag settings #}
{% set tag="AcademicObservatory" %}
tags:
- name: {{ tag }}
  description: The Academic Observatory API
```

jdddog commented on academic_observatory_api/openapi.yaml.jinja2:
```diff
@@ -0,0 +1,313 @@
+    get:
+      operationId: {{ operation_id_namespace }}pit_id_agg
+      tags:
+        - {{ reports_tag }}
```
```suggestion
        - {{ tag }}
```

jdddog commented on academic_observatory_api/openapi.yaml.jinja2:
```diff
@@ -0,0 +1,313 @@
+    get:
+      operationId: {{ operation_id_namespace }}query_agg
+      tags:
+        - {{ reports_tag }}
```
```suggestion
        - {{ tag }}
```

jdddog commented on academic_observatory_api/openapi.yaml.jinja2:
```diff
@@ -0,0 +1,313 @@
+    get:
+      operationId: {{ operation_id_namespace }}pit_id_subagg
+      tags:
+        - {{ reports_tag }}
```
```suggestion
        - {{ tag }}
```

jdddog commented on academic_observatory_api/openapi.yaml.jinja2:
```diff
@@ -0,0 +1,313 @@
+    get:
+      operationId: {{ operation_id_namespace }}query_subagg
+      tags:
+        - {{ reports_tag }}
```
```suggestion
        - {{ tag }}
```

jdddog reviewed (changes requested):
Hey Aniek, the changes are looking good, the feedback is mostly related to removing things related to the Observatory API.

jdddog commented on docs/deploy_terraform.md:
```diff
@@ -0,0 +1,114 @@
+  "id" : "random_id",
+  "name" : "my-dev-api-key",
+  "api_key" : "random_api_key"
+}
```
```suggestion
{
  "id" : "random_id",
  "name" : "my-dev-api-key",
  "api_key" : "random_api_key",
  "encoded" : "random_id:random_api_key base64 encoded"
}
```

jdddog commented on docs/deploy_terraform.md:
````diff
@@ -0,0 +1,114 @@
+}
+```
+
+Concat id:api_key and base64 encode (this final value is what is used for the Terraform variable 
+'elasticsearch_api_key'):
+```bash
+printf 'random_id:random_api_key' | base64
+```
````
Elastic v8 now returns this base64 encoded value directly.

aroelo replied:
Oooh, nice!

jdddog commented on docs/deploy_terraform.md:
````diff
@@ -0,0 +1,114 @@
+credentials = json.dumps(data)
+```
+
+Copy the 'credentials' value and use it for the 'google_cloud' Terraform variable, see below.
````
Terraform recommends setting GOOGLE_APPLICATION_CREDENTIALS as a secret environment variable to store the GCP credentials with the cloud backend: https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/getting_started#using-terraform-cloud-as-the-backend

This also means that the google_cloud variable doesn't need to be a secret.

jdddog commented on docs/deploy_terraform.md:
```diff
@@ -0,0 +1,114 @@
+
+| Variable      | Example                                                                                                                                                                                                               | HCL | Sensitive |
+|---------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----|-----------|
+| google_cloud  | {<br/>"project_id"="my-project-id",<br/>"credentials"="json-credentials",<br/>"region="us-central1"<br/>}                                                                                                             | Yes | Yes       |
```
The google_cloud example has a typo, missing " after region.

We can remove "credentials"="json-credentials" if we use the GOOGLE_APPLICATION_CREDENTIALS environment variable as above.

jdddog commented on docs/deploy_terraform.md:
```diff
@@ -0,0 +1,114 @@
+| google_cloud  | {<br/>"project_id"="my-project-id",<br/>"credentials"="json-credentials",<br/>"region="us-central1"<br/>}                                                                                                             | Yes | Yes       |
+| name          | ao                                                                                                                                                                                                                    | No  | No        |
+| domain_name   | my-project-id.ao.api.observatory.academy                                                                                                                                                                              | No  | Yes       |
+| backend_image | us-docker.pkg.dev/your-project-name/observatory-platform/observatory-api:0.3.1                                                                                                                                        | No  | Yes       |
```
Domain name and backend image don't need to be sensitive.

jdddog commented on docs/deploy_terraform.md:
```diff
@@ -0,0 +1,114 @@
+| domain_name   | my-project-id.ao.api.observatory.academy                                                                                                                                                                              | No  | Yes       |
+| backend_image | us-docker.pkg.dev/your-project-name/observatory-platform/observatory-api:0.3.1                                                                                                                                        | No  | Yes       |
+| gateway_image | gcr.io/endpoints-release/endpoints-runtime-serverless:2                                                                                                                                                               | No  | No        |
+| api_type      | {<br/>"type"="data_api",<br/>"observatory_organization"="",<br/>"observatory_workspace"="",<br/>elasticsearch_host="https://my-project-id.es.us-west1.gcp.cloud.es.io:9243",<br/>elasticsearch_api_key="APIKEY"<br/>} | Yes | Yes       |
```
elasticsearch_host doesn't need to be sensitive.

jdddog commented on docs/deploy_terraform.md:
```diff
@@ -0,0 +1,114 @@
+# Deploy the API with Terraform
+## Set up custom domain name
```
We should also add the the step to generate the openapi template: `coki-api-base generate-openapi-spec ../academic_observatory_api/openapi.yaml.jinja2 ./openapi.yaml.tpl --usage-type cloud_endpoints`

jdddog commented on docs/deploy_terraform.md:
```diff
@@ -0,0 +1,114 @@
+# Deploy the API with Terraform
+## Set up custom domain name
+Register a custom domain that will be used for the API.
```
How do we use a Google generated domain rather than a custom domain?

aroelo replied:
You could probably use the name from the Cloud Run gateway instance. 

In the API module a mapping is made between the custom domain and this name (https://github.com/The-Academic-Observatory/terraform-google-api/blob/main/main.tf#L146-L158)

```yaml
# Create custom domain mapping for cloud run gateway
resource "google_cloud_run_domain_mapping" "default" {
  location = google_cloud_run_service.api-gateway.location
  name     = var.domain_name

  metadata {
    namespace = var.google_cloud.project_id
  }

  spec {
    route_name = google_cloud_run_service.api-gateway.name
  }
}
```

You could maybe make this optional depending on whether the custom_domain is given or not.

Additionally you would then use this gateway name for the host in the openapi config when creating the endpoints service:
```
resource "google_endpoints_service" "api" {
  project      = var.google_cloud.project_id
  service_name = var.domain_name
  openapi_config = templatefile("./openapi.yaml.tpl", {
    host            = var.domain_name <--- should then be the gateway name
    backend_address = google_cloud_run_service.api-backend.status[0].url
  })
}
```

I don't have time to test this and work it out, but I think that are the two things that need to be modified.

jdddog replied:
Thanks Aniek, I'll add it as a future task.

jdddog reviewed (commented):
Hey Aniek, here is some more feedback after I've deployed it.

It would also be good if there were instructions on how to use coki-api-base to generate the Docker file and openapi files, perhaps that is better in coki-api-base though.

jdddog commented on terraform/main.tf:
```diff
@@ -0,0 +1,76 @@
+    workspaces {
+      prefix = "observatory-"
+    }
+  }
```
Terraform seemed to have updated the config for this to something like:

```
terraform {
  cloud {
    workspaces {
      prefix = "academic-observatory-api-"
    }
  }
}
```

Commit:
Update openapi.yaml.jinja2 and re-run openapi generator

Commit:
Update openapi.yaml.jinja2 and re-run openapi generator

Commit:
- Add step to create the OpenAPI template
- Update returned Elasticsearch key
- Update info to use google credentials environment variable
- Update info to use elasticsearch_host and elasticsearch_api_key variables directly

Commit:
- Use google credentials environment variable instead
- Use elasticsearch_host and elasticsearch_api_key variables directly

Commit:
Update workflow with 'cloud' block terraform settings

Commit:
Update workflow with 'cloud' block terraform settings

Commit:
Update workflow with 'cloud' block terraform settings

Commit:
Update workflow with 'cloud' block terraform settings

Commit:
Update workflow with 'cloud' block terraform settings

Commit:
Fix typo

Commit:
Fix github workflow

Commit:
Fix github workflow

Commit:
Use API module from registry

Commit:
Format table

Commit:
Add info on docker build

jdddog commented on terraform/variables.tf:
```diff
@@ -0,0 +1,49 @@
+variable "domain_name" {
+  description = "The custom domain name for the API, used for the google cloud endpoints service"
+  type        = string
+  sensitive   = true
```
```suggestion
  type        = string
```

jdddog commented on terraform/variables.tf:
```diff
@@ -0,0 +1,49 @@
+variable "backend_image" {
+  description = "The image URL that will be used for the Cloud Run backend, e.g. 'us-docker.pkg.dev/your-project-name/observatory-platform/observatory-api:0.3.1'"
+  type        = string
+  sensitive   = true
```
```suggestion
  type        = string
```

jdddog commented on terraform/variables.tf:
```diff
@@ -0,0 +1,49 @@
+variable "elasticsearch_api_key" {
+  description = "An API key for the Elasticsearch server"
+  type        = string
+}
```
```suggestion
  type        = string
  sensitive   = true
}
```

jdddog commented on terraform/variables.tf:
```diff
@@ -0,0 +1,49 @@
+    project_id = string
+    region     = string
+  })
+  sensitive = true
```
```suggestion
```

jdddog commented on terraform/outputs.tf:
```diff
@@ -0,0 +1,11 @@
+output "api-domain-name" {
+  value       = module.api.api-domain-name
+  description = "Custom domain name for the API"
+  sensitive   = true
+}
+
+output "api-gateway-url" {
+  value       = module.api.api-gateway-url
+  description = "Cloud run gateway URL for the API"
+  sensitive   = true
+}
```
Do these need to be sensitive? I think it would be useful to see them. It would need updating in the Terraform module too.

aroelo replied:
I don't think they need to be, was just being a bit conservative :-) I will update them to not be sensitive.

jdddog reviewed (changes requested):
Thanks Aniek, these changes are looking good, just a few small comments.

Commit:
Update 'sensitive' settings for terraform variables/outputs
